File: src/libs/sdk/dotnet/Ums.Sdk.Authorization/AuthorizationValidator.cs
using Ums.Sdk.Contracts;

namespace Ums.Sdk.Authorization;

/// <summary>
/// Reference implementation of <see cref="IAuthorizationValidator"/>.
/// Enforces the rules documented in <c>docs/domain/identity/auth-graph.md §3</c>:
/// Override over Template, Deny over Allow, NotGranted on missing entries, Expired on past validUntil,
/// SchemaUnsupported on MAJOR mismatch, SchemaMissing when schemaVersion is absent.
/// </summary>
public sealed class AuthorizationValidator : IAuthorizationValidator
{
    private const string PrimScope    = "RequiresScope";
    private const string PrimMenu     = "RequiresMenuOption";
    private const string PrimDomain   = "RequiresDomainAccess";
    private const string PrimFlag     = "RequiresFeatureFlag";

    public AuthorizationDecision RequireScope(AuthorizationGraph? graph, string scope)
    {
        var pre = PreCheck(graph, PrimScope, scope);
        if (pre is not null) return pre;

        // Deny precedence: a deny anywhere in menuAccess or domainPermissions matching this scope
        // beats an Allow. The scope is reconstructed as "{code}.{actionCode}" — we check both surfaces.
        var (resource, action) = ParseScope(scope);
        if (resource is null || action is null)
        {
            return AuthorizationDecision.NotGranted(PrimScope, scope,
                UmsErrorCodes.ScopeNotGranted,
                $"Scope '{scope}' is not in the canonical 'RESOURCE.ACTION' format.",
                graphRequestId: null, validUntil: graph!.ValidUntil);
        }

        if (IsDeniedInGraph(graph!, resource, action))
        {
            return AuthorizationDecision.Deny(PrimScope, scope,
                UmsErrorCodes.ScopeDenied,
                $"Scope '{scope}' is explicitly denied in the authorization graph.",
                graphRequestId: null, validUntil: graph!.ValidUntil);
        }

        return graph!.Scopes.Contains(scope, StringComparer.Ordinal)
            ? AuthorizationDecision.Granted(PrimScope, scope, graphRequestId: null, validUntil: graph.ValidUntil)
            : AuthorizationDecision.NotGranted(PrimScope, scope,
                UmsErrorCodes.ScopeNotGranted,
                $"Scope '{scope}' is not present in the authorization graph.",
                graphRequestId: null, validUntil: graph.ValidUntil);
    }

    public AuthorizationDecision RequireMenuOption(AuthorizationGraph? graph, string optionCode)
    {
        var pre = PreCheck(graph, PrimMenu, optionCode);
        if (pre is not null) return pre;

        foreach (var module in graph!.MenuAccess)
        foreach (var menu in module.Menus)
        foreach (var sub in menu.SubMenus)
        foreach (var opt in sub.Options)
        {
            if (!string.Equals(opt.Code, optionCode, StringComparison.Ordinal)) continue;
            return opt.Effect switch
            {
                AccessEffect.Allow => AuthorizationDecision.Granted(PrimMenu, optionCode,
                    validUntil: graph.ValidUntil),
                AccessEffect.Deny => AuthorizationDecision.Deny(PrimMenu, optionCode,
                    UmsErrorCodes.MenuOptionDenied,
                    $"Menu option '{optionCode}' is explicitly denied (source: {opt.Source}).",
                    validUntil: graph.ValidUntil),
                _ => AuthorizationDecision.NotGranted(PrimMenu, optionCode,
                    UmsErrorCodes.MenuOptionNotGranted,
                    $"Menu option '{optionCode}' resolves to NotGranted.",
                    validUntil: graph.ValidUntil)
            };
        }

        return AuthorizationDecision.NotGranted(PrimMenu, optionCode,
            UmsErrorCodes.MenuOptionNotGranted,
            $"Menu option '{optionCode}' is not present in the authorization graph.",
            validUntil: graph.ValidUntil);
    }

    public AuthorizationDecision RequireDomainAccess(AuthorizationGraph? graph, string resourceCode, string actionCode)
    {
        var target = $"{resourceCode}.{actionCode}";
        var pre = PreCheck(graph, PrimDomain, target);
        if (pre is not null) return pre;

        foreach (var dr in graph!.DomainPermissions)
        {
            if (!string.Equals(dr.Resource.Code, resourceCode, StringComparison.Ordinal)) continue;
            foreach (var act in dr.Actions)
            {
                if (!string.Equals(act.ActionCode, actionCode, StringComparison.Ordinal)) continue;
                return act.Effect switch
                {
                    AccessEffect.Allow => AuthorizationDecision.Granted(PrimDomain, target,
                        validUntil: graph.ValidUntil),
                    AccessEffect.Deny => AuthorizationDecision.Deny(PrimDomain, target,
                        UmsErrorCodes.DomainAccessDenied,
                        $"Domain access '{target}' is explicitly denied (source: {act.Source}).",
                        validUntil: graph.ValidUntil),
                    _ => AuthorizationDecision.NotGranted(PrimDomain, target,
                        UmsErrorCodes.DomainAccessNotGranted,
                        $"Domain access '{target}' resolves to NotGranted.",
                        validUntil: graph.ValidUntil)
                };
            }
        }

        return AuthorizationDecision.NotGranted(PrimDomain, target,
            UmsErrorCodes.DomainAccessNotGranted,
            $"Domain access '{target}' is not present in the authorization graph.",
            validUntil: graph.ValidUntil);
    }

    public AuthorizationDecision RequireFeatureFlag(AuthorizationGraph? graph, string flagCode)
    {
        var pre = PreCheck(graph, PrimFlag, flagCode);
        if (pre is not null) return pre;

        var flag = graph!.FeatureFlags.FirstOrDefault(
            f => string.Equals(f.FlagCode, flagCode, StringComparison.Ordinal));

        if (flag is null)
        {
            return AuthorizationDecision.NotGranted(PrimFlag, flagCode,
                UmsErrorCodes.FeatureFlagNotFound,
                $"Feature flag '{flagCode}' is not present in the authorization graph.",
                validUntil: graph.ValidUntil);
        }

        return flag.IsEnabled
            ? AuthorizationDecision.Granted(PrimFlag, flagCode, validUntil: graph.ValidUntil)
            : AuthorizationDecision.NotGranted(PrimFlag, flagCode,
                UmsErrorCodes.FeatureFlagDisabled,
                $"Feature flag '{flagCode}' is present but isEnabled is false.",
                validUntil: graph.ValidUntil);
    }

    public AuthorizationDecision AssertTenant(AuthorizationGraph? graph, string expectedTenantCode)
    {
        var pre = PreCheck(graph, "AssertTenant", expectedTenantCode);
        if (pre is not null) return pre;

        return string.Equals(graph!.Context.Tenant.Code, expectedTenantCode, StringComparison.Ordinal)
            ? AuthorizationDecision.Granted("AssertTenant", expectedTenantCode, validUntil: graph.ValidUntil)
            : AuthorizationDecision.TenantMismatch(expectedTenantCode, graph.Context.Tenant.Code);
    }

    // --- Helpers ---------------------------------------------------------

    private static AuthorizationDecision? PreCheck(AuthorizationGraph? graph, string primitive, string target)
    {
        if (graph is null)
            return AuthorizationDecision.GraphMissing(primitive, target);
        if (string.IsNullOrWhiteSpace(graph.SchemaVersion))
            return AuthorizationDecision.SchemaMissing(primitive, target);
        if (!SchemaVersion.IsSupported(graph.SchemaVersion))
            return AuthorizationDecision.SchemaUnsupported(primitive, target, graph.SchemaVersion);
        if (graph.ValidUntil <= DateTimeOffset.UtcNow)
            return AuthorizationDecision.Expired(primitive, target, graph.ValidUntil);
        return null;
    }

    private static (string? Resource, string? Action) ParseScope(string scope)
    {
        var idx = scope.IndexOf('.');
        if (idx <= 0 || idx == scope.Length - 1) return (null, null);
        return (scope[..idx], scope[(idx + 1)..]);
    }

    private static bool IsDeniedInGraph(AuthorizationGraph graph, string resourceCode, string actionCode)
    {
        // Probe menuAccess: any option whose code equals the resource (UI surface) AND actionCode matches.
        foreach (var module in graph.MenuAccess)
        foreach (var menu in module.Menus)
        foreach (var sub in menu.SubMenus)
        foreach (var opt in sub.Options)
        {
            if (opt.Effect != AccessEffect.Deny) continue;
            if (string.Equals(opt.Code, resourceCode, StringComparison.Ordinal) &&
                string.Equals(opt.ActionCode, actionCode, StringComparison.Ordinal))
                return true;
        }

        // Probe domainPermissions.
        foreach (var dr in graph.DomainPermissions)
        {
            if (!string.Equals(dr.Resource.Code, resourceCode, StringComparison.Ordinal)) continue;
            foreach (var act in dr.Actions)
            {
                if (act.Effect == AccessEffect.Deny &&
                    string.Equals(act.ActionCode, actionCode, StringComparison.Ordinal))
                    return true;
            }
        }

        return false;
    }
}

File: src/libs/sdk/typescript/sdk-authorization/src/authorization-validator.ts
import type { AuthorizationGraph } from '@ums/sdk-contracts';
import { isSchemaVersionSupported } from '@ums/sdk-contracts';
import { UmsErrorCodes } from '@ums/sdk-contracts';
import { Decisions, type AuthorizationDecision } from './authorization-decision.js';

const PrimScope = 'RequiresScope';
const PrimMenu = 'RequiresMenuOption';
const PrimDomain = 'RequiresDomainAccess';
const PrimFlag = 'RequiresFeatureFlag';

/**
 * Pure validator. Same semantics as `Ums.Sdk.Authorization.AuthorizationValidator` (.NET) and
 * the NestJS guard — given a graph and a probe, returns a decision. No I/O, no framework deps.
 *
 * Enforces (in pre-check order): graph presence → schema version present → schema in range → not
 * expired. Then per-primitive: deny precedence over allow, override over template, missing entries
 * yield NotGranted.
 */
export class AuthorizationValidator {
  requireScope(graph: AuthorizationGraph | null, scope: string): AuthorizationDecision {
    const pre = this.preCheck(graph, PrimScope, scope);
    if (pre !== null) return pre;

    const parsed = parseScope(scope);
    if (parsed === null) {
      return Decisions.notGranted(
        PrimScope,
        scope,
        UmsErrorCodes.ScopeNotGranted,
        `Scope '${scope}' is not in the canonical 'RESOURCE.ACTION' format.`,
        graph!.validUntil
      );
    }

    if (isDeniedInGraph(graph!, parsed.resource, parsed.action)) {
      return Decisions.deny(
        PrimScope,
        scope,
        UmsErrorCodes.ScopeDenied,
        `Scope '${scope}' is explicitly denied in the authorization graph.`,
        graph!.validUntil
      );
    }

    return graph!.scopes.includes(scope)
      ? Decisions.granted(PrimScope, scope, graph!.validUntil)
      : Decisions.notGranted(
          PrimScope,
          scope,
          UmsErrorCodes.ScopeNotGranted,
          `Scope '${scope}' is not present in the authorization graph.`,
          graph!.validUntil
        );
  }

  requireMenuOption(graph: AuthorizationGraph | null, optionCode: string): AuthorizationDecision {
    const pre = this.preCheck(graph, PrimMenu, optionCode);
    if (pre !== null) return pre;

    for (const module of graph!.menuAccess) {
      for (const menu of module.menus) {
        for (const sub of menu.subMenus) {
          for (const opt of sub.options) {
            if (opt.code !== optionCode) continue;
            switch (opt.effect) {
              case 'Allow':
                return Decisions.granted(PrimMenu, optionCode, graph!.validUntil);
              case 'Deny':
                return Decisions.deny(
                  PrimMenu,
                  optionCode,
                  UmsErrorCodes.MenuOptionDenied,
                  `Menu option '${optionCode}' is explicitly denied (source: ${opt.source}).`,
                  graph!.validUntil
                );
              default:
                return Decisions.notGranted(
                  PrimMenu,
                  optionCode,
                  UmsErrorCodes.MenuOptionNotGranted,
                  `Menu option '${optionCode}' resolves to NotGranted.`,
                  graph!.validUntil
                );
            }
          }
        }
      }
    }

    return Decisions.notGranted(
      PrimMenu,
      optionCode,
      UmsErrorCodes.MenuOptionNotGranted,
      `Menu option '${optionCode}' is not present in the authorization graph.`,
      graph!.validUntil
    );
  }

  requireDomainAccess(
    graph: AuthorizationGraph | null,
    resourceCode: string,
    actionCode: string
  ): AuthorizationDecision {
    const target = `${resourceCode}.${actionCode}`;
    const pre = this.preCheck(graph, PrimDomain, target);
    if (pre !== null) return pre;

    for (const dr of graph!.domainPermissions) {
      if (dr.resource.code !== resourceCode) continue;
      for (const act of dr.actions) {
        if (act.actionCode !== actionCode) continue;
        switch (act.effect) {
          case 'Allow':
            return Decisions.granted(PrimDomain, target, graph!.validUntil);
          case 'Deny':
            return Decisions.deny(
              PrimDomain,
              target,
              UmsErrorCodes.DomainAccessDenied,
              `Domain access '${target}' is explicitly denied (source: ${act.source}).`,
              graph!.validUntil
            );
          default:
            return Decisions.notGranted(
              PrimDomain,
              target,
              UmsErrorCodes.DomainAccessNotGranted,
              `Domain access '${target}' resolves to NotGranted.`,
              graph!.validUntil
            );
        }
      }
    }

    return Decisions.notGranted(
      PrimDomain,
      target,
      UmsErrorCodes.DomainAccessNotGranted,
      `Domain access '${target}' is not present in the authorization graph.`,
      graph!.validUntil
    );
  }

  requireFeatureFlag(graph: AuthorizationGraph | null, flagCode: string): AuthorizationDecision {
    const pre = this.preCheck(graph, PrimFlag, flagCode);
    if (pre !== null) return pre;

    const flag = graph!.featureFlags.find((f) => f.flagCode === flagCode);
    if (flag === undefined) {
      return Decisions.notGranted(
        PrimFlag,
        flagCode,
        UmsErrorCodes.FeatureFlagNotFound,
        `Feature flag '${flagCode}' is not present in the authorization graph.`,
        graph!.validUntil
      );
    }
    return flag.isEnabled
      ? Decisions.granted(PrimFlag, flagCode, graph!.validUntil)
      : Decisions.notGranted(
          PrimFlag,
          flagCode,
          UmsErrorCodes.FeatureFlagDisabled,
          `Feature flag '${flagCode}' is present but isEnabled is false.`,
          graph!.validUntil
        );
  }

  assertTenant(graph: AuthorizationGraph | null, expectedTenantCode: string): AuthorizationDecision {
    const pre = this.preCheck(graph, 'AssertTenant', expectedTenantCode);
    if (pre !== null) return pre;

    return graph!.context.tenant.code === expectedTenantCode
      ? Decisions.granted('AssertTenant', expectedTenantCode, graph!.validUntil)
      : Decisions.tenantMismatch(expectedTenantCode, graph!.context.tenant.code);
  }

  private preCheck(graph: AuthorizationGraph | null, primitive: string, target: string): AuthorizationDecision | null {
    if (graph === null) return Decisions.graphMissing(primitive, target);
    if (graph.schemaVersion === undefined || graph.schemaVersion === null || graph.schemaVersion === '') {
      return Decisions.schemaMissing(primitive, target);
    }
    if (!isSchemaVersionSupported(graph.schemaVersion)) {
      return Decisions.schemaUnsupported(primitive, target, graph.schemaVersion);
    }
    if (new Date(graph.validUntil).getTime() <= Date.now()) {
      return Decisions.expired(primitive, target, graph.validUntil);
    }
    return null;
  }
}

function parseScope(scope: string): { resource: string; action: string } | null {
  const idx = scope.indexOf('.');
  if (idx <= 0 || idx === scope.length - 1) return null;
  return { resource: scope.substring(0, idx), action: scope.substring(idx + 1) };
}

function isDeniedInGraph(graph: AuthorizationGraph, resourceCode: string, actionCode: string): boolean {
  for (const module of graph.menuAccess) {
    for (const menu of module.menus) {
      for (const sub of menu.subMenus) {
        for (const opt of sub.options) {
          if (opt.effect === 'Deny' && opt.code === resourceCode && opt.actionCode === actionCode) {
            return true;
          }
        }
      }
    }
  }
  for (const dr of graph.domainPermissions) {
    if (dr.resource.code !== resourceCode) continue;
    for (const act of dr.actions) {
      if (act.effect === 'Deny' && act.actionCode === actionCode) return true;
    }
  }
  return false;
}
